# Проект "Транзакции в онлайн-игре"

## Описание проекта

### Задачи проекта

- Проанализировать транзакции в онлайн-игре.
- Выяснить, какая доля игроков покупает внутриигровую валюту «райские лепестки» за реальные деньги.
- Узнать, есть ли зависимость доли платящих игроков от расы персонажа.
- Детально изучить, как происходит покупка эпических предметов внутри игры.
- Проверить гипотезу о том, что игра за некоторые расы сложнее и требует большего количества покупок эпических предметов, которые помогают в прохождении.
- Сформулировать рекомендации для разработчиков для улучшения внутриигрового баланса.

### Описание данных

#### Таблица: users

Хранит данные об игроках, их характеристиках, демографии и статусе платежей.

- `id` — идентификатор игрока (первичный ключ).  
- `tech_nickname` — технический никнейм игрока.  
- `class_id` — идентификатор класса персонажа (внешний ключ к таблице `classes`).  
- `ch_id` — идентификатор легендарного умения (внешний ключ к таблице `skills`).  
- `birthdate` — дата рождения персонажа.  
- `pers_gender` — пол персонажа.  
- `registration_dt` — дата регистрации пользователя в игре.  
- `server` — сервер, на котором играет пользователь.  
- `race_id` — идентификатор расы персонажа (внешний ключ к таблице `race`).  
- `payer` — признак платного пользователя: 1 — платящий (покупал «райские лепестки»), 0 — не платящий.  
- `loc_id` — идентификатор страны игрока (внешний ключ к таблице `country`).  

---

#### Таблица: events

Содержит историю покупок внутриигровых предметов.

- `transaction_id` — идентификатор транзакции (первичный ключ).  
- `id` — идентификатор игрока (внешний ключ к таблице `users`).  
- `date` — дата совершения покупки.  
- `time` — время совершения покупки.  
- `item_code` — код эпического предмета (внешний ключ к таблице `items`).  
- `amount` — стоимость покупки в валюте «райские лепестки».  
- `seller_id` — идентификатор продавца предмета.  

---

#### Таблица: skills

Справочник легендарных умений персонажей.

- `ch_id` — идентификатор легендарного умения (первичный ключ).  
- `legendary_skill` — название легендарного умения.  

---

#### Таблица: race

Справочник рас персонажей.

- `race_id` — идентификатор расы (первичный ключ).  
- `race` — название расы персонажа.  

---

#### Таблица: country

Справочник стран, к которым привязаны игроки.

- `loc_id` — идентификатор страны (первичный ключ).  
- `location` — название страны.  

---

#### Таблица: classes

Справочник классов персонажей.

- `class_id` — идентификатор класса (первичный ключ).  
- `class` — название класса персонажа.  

---

#### Таблица: items

Справочник эпических предметов.

- `item_code` — код эпического предмета (первичный ключ).  
- `game_items` — название эпического предмета.  

### Инструменты анализа

Для анализа данных был использован PostgreSQL со следующими функциями и техниками:
- Common Table Expressions (CTE) — конструкции WITH AS, в том числе с несколькими последовательными CTE в одном запросе, для пошаговой агрегации данных перед финальным расчетом
- Соединение таблиц — JOIN для объединения таблиц
- Агрегатные функции — COUNT, COUNT(DISTINCT), SUM, MIN, MAX, AVG
- Условная агрегация — COUNT(CASE WHEN THEN END) для подсчета по определенному признаку внутри общей группировки
- Статистические функции распределения — PERCENTILE_DISC(0.5) WITHIN GROUP для расчёта медианы и STDDEV для расчёта стандартного отклонения
- Приведение типов — явное приведение через ::NUMERIC для получения корректных дробных результатов при делении
- Округление — ROUND для приведения долей и средних значений к читаемому виду
- Группировка и сортировка — GROUP BY и ORDER BY для агрегации и ранжирования результатов
- Подзапросы — вложенный SELECT внутри вычисляемого поля
- Ограничение выборки — LIMIT для отображения топ-N результатов

## Работа над проектом

### Настройка работы с SQL

In [1]:
%load_ext sql

In [2]:
!pip install psycopg2-binary

In [3]:
!pip install "prettytable<3.0.0"

In [4]:
import psycopg2

### Выгрузка данных

In [5]:
import urllib.parse

user = "praktikum_student"
password = "Sdf4$2;d-d30pp"
host = "rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net"
port = 6432
dbname = "data-analyst-fantasy"

encoded_password = urllib.parse.quote_plus(password)
conn_str = f"postgresql+psycopg2://{user}:{encoded_password}@{host}:{port}/{dbname}?sslmode=require"

%sql $conn_str

### Анализ данных

In [6]:
%%sql
-- Доля платящих пользователей по всем данным:

WITH paying_users AS (  
    SELECT
        COUNT(*) AS total_users,
        COUNT(DISTINCT CASE WHEN payer = 1 THEN id END) AS total_paying_users
    FROM fantasy.users u
)
SELECT
    *,
    ROUND(total_paying_users / total_users::NUMERIC, 4) AS paying_users_share
FROM paying_users;

 * postgresql+psycopg2://praktikum_student:***@rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net:6432/data-analyst-fantasy?sslmode=require
1 rows affected.


total_users,total_paying_users,paying_users_share
22214,3929,0.1769


Вывод:
- Процент платящих игроков для всей игры составляет 17,69% – то есть примерно каждый 6 игрок осуществляет покупки за реальные деньги.

In [7]:
%%sql
-- Доля платящих пользователей в разрезе расы персонажа:
    
WITH user_race_data AS (
    SELECT
        r.race,
        COUNT(*) AS total_users,
        COUNT(CASE WHEN u.payer = 1 THEN u.id END) AS total_paying_users
    FROM fantasy.users u
    JOIN fantasy.race r ON u.race_id = r.race_id
    GROUP BY r.race
)
SELECT
    race,
    total_paying_users,
    total_users,
    ROUND(total_paying_users::NUMERIC / total_users, 4) AS paying_users_share
FROM user_race_data
ORDER BY paying_users_share DESC;

 * postgresql+psycopg2://praktikum_student:***@rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net:6432/data-analyst-fantasy?sslmode=require
7 rows affected.


race,total_paying_users,total_users,paying_users_share
Demon,238,1229,0.1937
Hobbit,659,3648,0.1806
Human,1114,6328,0.1760
Orc,636,3619,0.1757
Northman,626,3562,0.1757
Angel,229,1327,0.1726
Elf,427,2501,0.1707


Вывод:
- Среди расы демонов процент платящих игроков заметно превышает среднее значение.
- Среди других рас размах этого показателя не превышает 1%.

In [8]:
%%sql
-- Статистические показатели по полю amount:

SELECT
    COUNT(*) AS total_purchases,
    SUM(amount) AS purchase_amount,
    MIN(amount) AS min_purchase,
    MAX(amount) AS max_purchase,
    ROUND(AVG(amount)::numeric, 2) AS avg_purchase,
    PERCENTILE_DISC(0.5) WITHIN GROUP(ORDER BY amount) AS median_purchase,
    ROUND(STDDEV(amount)::NUMERIC, 2) AS std_dev_purchase
FROM fantasy.events;

 * postgresql+psycopg2://praktikum_student:***@rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net:6432/data-analyst-fantasy?sslmode=require
1 rows affected.


total_purchases,purchase_amount,min_purchase,max_purchase,avg_purchase,median_purchase,std_dev_purchase
1307678,686615040.0,0.0,486615.1,525.69,74.86,2517.35


Вывод:
- Всего совершено свыше 1,3 млн игровых покупок.
- Минимальное значение стоимости - 0, максимальное - 486615,1.
- Среднее значение в 7 раз превышает медианное, следовательно присутствуют очень дорогие покупки, которые искажают среднее значение.
- Разброс данных составляет примерно 2500 “райских лепестков”, что весьма много относительно среднего и медианного значений.

In [9]:
%%sql
-- Аномальные нулевые покупки:
    
WITH zero_purchases AS (
    SELECT COUNT(*) as count
    FROM fantasy.events
    WHERE amount = 0
)
SELECT 
    zp.count AS total_zero_purchases,
    ROUND(zp.count::NUMERIC / (SELECT COUNT(*) FROM fantasy.events), 4) AS zero_purchases_share
FROM zero_purchases zp;

 * postgresql+psycopg2://praktikum_student:***@rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net:6432/data-analyst-fantasy?sslmode=require
1 rows affected.


total_zero_purchases,zero_purchases_share
907,0.0007


Вывод:
- Присутствуют покупки с нулевой стоимостью, их доля от общего числа покупок около 0,07%, что не особо влияет на общий результат.

In [10]:
%%sql
-- Популярные эпические предметы:
    
WITH total_stats AS (
    SELECT
        COUNT(*) AS total_purchases_all,
        COUNT(DISTINCT id) AS total_paying_users
    FROM fantasy.events
    WHERE amount <> 0
),
item_stats AS (
    SELECT
        i.game_items,
        COUNT(*) AS total_purchases,
        COUNT(DISTINCT e.id) AS unique_buyers
    FROM fantasy.events e
    JOIN fantasy.items i USING(item_code)
    WHERE amount <> 0
    GROUP BY i.game_items
)
SELECT
    game_items,
    total_purchases,
    ROUND(total_purchases::numeric / total_purchases_all, 4) AS percentage_share,
    ROUND(unique_buyers::numeric / total_paying_users, 4) AS percentage_paying_users
FROM item_stats, total_stats
ORDER BY total_purchases DESC
LIMIT 10;

 * postgresql+psycopg2://praktikum_student:***@rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net:6432/data-analyst-fantasy?sslmode=require
10 rows affected.


game_items,total_purchases,percentage_share,percentage_paying_users
Book of Legends,1004516,0.7687,0.8841
Bag of Holding,271875,0.2081,0.8677
Necklace of Wisdom,13828,0.0106,0.1180
Gems of Insight,3833,0.0029,0.0671
Treasure Map,3183,0.0024,0.0594
Amulet of Protection,1078,0.0008,0.0323
Silver Flask,795,0.0006,0.0459
Strength Elixir,580,0.0004,0.0240
Glowing Pendant,563,0.0004,0.0257
Gauntlets of Might,514,0.0004,0.0204


Вывод:
- Среди эпических предметов особо выделяется “Book of Legends”, продажи этого предмета составляют 76,87% от общего числа продаж.
- Также  высокие продажи у “Bag of Holding” - 20,81%.
- При этом оба предмета хотя бы раз покупали 87-88% всех игроков.

In [11]:
%%sql
-- Анализ зависимости активности игроков от расы персонажа:
    
WITH user_data AS (
    SELECT
        r.race,
        u.id,
        u.payer,
        e.transaction_id,
        e.amount
    FROM fantasy.users u
    JOIN fantasy.race r USING(race_id)
    LEFT JOIN fantasy.events e ON u.id = e.id AND e.amount <> 0
),
aggregated_data AS (
    SELECT
        race,
        COUNT(DISTINCT id) AS total_users,
        COUNT(DISTINCT CASE WHEN payer = 1 AND transaction_id IS NOT NULL THEN id END) AS paying_users,
        COUNT(DISTINCT CASE WHEN transaction_id IS NOT NULL THEN id END) AS purchasing_users,
        COUNT(transaction_id) AS total_purchases,
        SUM(amount) AS sum_purchases
    FROM user_data
    GROUP BY race
)
SELECT
    race,
    total_users,
    purchasing_users,
    ROUND(purchasing_users::NUMERIC / total_users, 4) AS purchasing_users_share,
    ROUND(paying_users::NUMERIC / purchasing_users, 4) AS paying_users_share_among_purchasing,
    ROUND(total_purchases::NUMERIC / purchasing_users, 0) AS avg_purchases,
    ROUND(sum_purchases::NUMERIC / total_purchases, 2) AS avg_cost,
    ROUND(sum_purchases::NUMERIC / purchasing_users, 2) AS avg_total_cost
FROM aggregated_data
ORDER BY purchasing_users DESC;

 * postgresql+psycopg2://praktikum_student:***@rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net:6432/data-analyst-fantasy?sslmode=require
7 rows affected.


race,total_users,purchasing_users,purchasing_users_share,paying_users_share_among_purchasing,avg_purchases,avg_cost,avg_total_cost
Human,6328,3921,0.6196,0.1801,121,403.05,48931.65
Orc,3619,2276,0.6289,0.1740,82,510.92,41761.56
Hobbit,3648,2266,0.6212,0.1770,86,552.91,47621.80
Northman,3562,2229,0.6258,0.1821,82,761.52,62522.21
Elf,2501,1543,0.6170,0.1627,79,682.32,53760.60
Angel,1327,820,0.6179,0.1671,107,455.64,48665.00
Demon,1229,737,0.5997,0.1995,78,529.02,41194.71


Вывод:
- Доля покупателей редких предметов среди всех рас примерно одинакова, как и доля игроков, покупающих игровую валюту за реальные деньги.
- Однако при этом средняя сумма покупок на каждого игрока у некоторых рас значительно выше - так, северянин тратит в среднем на 20 тыс. лепестков больше, чем представитель демонов или орков.
- Следовательно гипотеза подтверждается частично - прохождение игры за разные расы требует примерно равного количества покупок, но при этом совсем разную сумму игровой валюты.

## Итоги

### Итоговые выводы
- Доля платящих игроков по всей игре составляет 17,69% — то есть примерно каждый шестой игрок хотя бы раз покупал внутриигровую валюту за реальные деньги.
- В разрезе рас доля платящих игроков в целом однородна, за исключением расы демонов — у неё этот показатель заметно выше среднего, у остальных рас разброс не превышает 1%.
- Всего в игре зафиксировано более 1,3 млн покупок. Среднее значение стоимости покупки в 7 раз превышает медианное — значит, есть небольшое число очень дорогих покупок, которые сильно искажают среднюю картину. Также присутствует заметный разброс (стандартное отклонение) относительно среднего и медианы.
- Покупки с нулевой стоимостью занимают незначительную долю (около 0,07%) и не влияют существенно на результаты анализа.
- Среди эпических предметов ярко выражена концентрация спроса: "Book of Legends" составляет почти 77% всех продаж, "Bag of Holding" — ещё около 21%. Оба предмета хотя бы раз покупали 87–88% всех игроков, то есть это фактически обязательные для прохождения предметы.
- Доля игроков, совершающих покупки редких предметов, и доля платящих реальными деньгами примерно одинаковы среди рас. Однако средняя сумма трат на одного игрока существенно различается: например, северяне тратят в среднем на 20 тыс. "лепестков" больше, чем демоны или орки.
- Таким образом, гипотеза о разной сложности прохождения подтвердилась лишь частично: количество необходимых покупок эпических предметов у рас сопоставимо, но стоимость прохождения в игровой валюте у разных рас заметно разная.

### Рекомендации

- Пересмотреть цены на эпические предметы в разрезе рас — вероятно, для рас с высокими средними тратами (например, северяне) те же предметы обходятся дороже или требуют больше повторных покупок; стоит выровнять стоимость или количество необходимых транзакций между расами.
- Изучить баланс характеристик персонажей по расам, из-за которого игрокам одних рас, возможно, требуется больше усиления через покупки, чтобы достичь того же игрового результата, что и представителям других рас.
- Обратить внимание на расу демонов как на аномалию по доле платящих игроков — проверить, не связано ли это с дисбалансом (слишком легкое или слишком тяжелое прохождение), стимулирующим повышенные траты.
- Проверить экономику ключевых предметов ("Book of Legends" и "Bag of Holding") — раз они являются почти обязательными для подавляющего большинства игроков, важно убедиться, что их цена одинаково ощутима для всех рас, а не создаёт разный порог входа.
- Ввести регулярный мониторинг метрики "средние траты на игрока по расе" после внесения изменений в баланс, чтобы отследить, действительно ли расхождение между расами сокращается.
- Рассмотреть возможность компенсационных механик (бонусные внутриигровые ресурсы, скидки на эпические предметы) для рас с наибольшими средними тратами, чтобы сократить разницу без резкого изменения общей игровой экономики.